src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/SamlSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                // throws if not found.
+                bool foundAudienceCondition = false;
+                if (_allowedAudienceUris == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLAudienceUrisNotFound)));
+                }
+
+                foreach (MSSaml.SamlCondition samlCondition in  samlToken.Assertion.Conditions.Conditions)
+                {
+                    if(samlCondition is MSSaml.SamlAudienceRestrictionCondition)
+                    {
+                        foundAudienceCondition = true;
+                        if (!ValidateAudienceRestriction((MSSaml.SamlAudienceRestrictionCondition)samlCondition))
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLAudienceUriValidationFailed)));
+                        }
+
+                    }
+                }
+
+                if (!foundAudienceCondition)
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLAudienceUriValidationFailed)));
+            }
+
+            return new ValueTask<ReadOnlyCollection<IAuthorizationPolicy>>(policies);
+        }
+
+        protected virtual bool ValidateAudienceRestriction(MSSaml.SamlAudienceRestrictionCondition audienceRestrictionCondition)
+        {
+            foreach (Uri audienceUri in audienceRestrictionCondition.Audiences)
+            {
+
+                for (int j = 0; j < _allowedAudienceUris.Count; j++)
+                {
+                    if (StringComparer.Ordinal.Compare(audienceUri.AbsoluteUri, _allowedAudienceUris[j]) == 0)
+                        return true;
+                    else if (Uri.IsWellFormedUriString(_allowedAudienceUris[j], UriKind.Absolute))
+                    {
+                        Uri uri = new Uri(_allowedAudienceUris[j]);
+                        if (audienceUri.Equals(uri))
+                            return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public virtual ClaimSet ResolveClaimSet(SecurityToken token, out ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies)
+        {
+            if (token == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(token));
+            authorizationPolicies = null;
+
+            for (int i = 0; i < _supportingAuthenticators.Count; ++i)
+            {
+                if (_supportingAuthenticators[i].CanValidateToken(token))
+                {
+                    authorizationPolicies = _supportingAuthenticators[i].ValidateToken(token);
+                    AuthorizationContext authContext = AuthorizationContext.CreateDefaultAuthorizationContext(authorizationPolicies);
```

Comment:
ValidateTokenAsync, await it, and make the method return a ValueTask and be async.